CorsixTH/Lua/humanoid_actions/seek_room.lua

```diff
@@ -246,115 +253,162 @@ local function action_seek_room_start(action, humanoid)
     return
   end
   -- Tries to find the room, if it is a diagnosis room, try to find any room in the diagnosis list.
-  local room = action_seek_room_find_room(action, humanoid)
+  if not humanoid.diagnosed or action.room_type == "research" then
+    local room = action_seek_room_find_room(action, humanoid)
+    -- if we have the room but not the staff we shouldn't seek out the room either
+    if room then
+      if humanoid.message then
+        TheApp.ui.bottom_panel:removeMessage(humanoid)
+      end
+      action_seek_room_goto_room(room, humanoid, action.diagnosis_room)
+      return
+    end
+  end
 
-  if room then
-    action_seek_room_goto_room(room, humanoid, action.diagnosis_room)
-  else
-    if not action.done_init then
-      action.done_init = true
-      action.must_happen = true
+  -- check we can treat the patient - and just shortcut the processing
+  local req = humanoid.hospital:checkDiseaseRequirements(humanoid.disease.id)
+  if humanoid.diagnosed and not req then
+    local room = action_seek_room_find_room(action, humanoid)
+    -- if we have the room but not the staff we shouldn't seek out the room either
+    if room then
+      if humanoid.message then
+        TheApp.ui.bottom_panel:removeMessage(humanoid)
+      end
+      action_seek_room_goto_room(room, humanoid, action.diagnosis_room)
+      return
+    end
+  end
```

Comment:
Similar logic for treatment as diagnosis, just checking if we alos have outstanding treatment requirements